### ISSUE TITLE
stack address

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/unsafe-code-guidelines#312]
Code sometimes has a need to get an address from the stack — often for debugging, but sometimes for other reasons. Consider this snippet of code from rayon-rs/rayon#571:

```rust
let base = 0;
let addr = &base as *const _ as u64;
```

the goal here is just to get "some stack pointer" to use as a (quite weak) seed for a PRNG. This feels... sketchy to me, I thought it was worth noting.

**Comment by RalfJung** [comment]
>  The question -- which I think is relevant outside of the question of how rayon seeds its PRNG -- is whether we guarantee that addr points on the stack.

On the Abstract Machine level, I don't think there is such a thing as a "stack", at least not as something that exists in addressable memory. So your question is ill-formed there.

If we want to have implementation-specific guarantees, we could probably say that the allocations that back local variables such as `base` are in "the stack" (referring to some platform-specific region of memory). Is there precedent for an explicit statement like this in other languages?

**Comment by Lokathor** [comment]
well you can have variables that are local to a function, and you can also reference those variables, and those references can be used in functions that you call, and many CPUs have a stack pointer register, but that's about all that can be said overall.

If you want to read the stack pointer, read the stack pointer with inline asm. Otherwise, the values don't have to go on the stack. Particularly, references to a constant could easily point into the binary's read-only data.

**Comment by thomcc** [comment]
It's possible in some OSes (linux, for example) for code to manually spawn a thread[^1], which involves passing the memory to be used for the thread's stack. Being able to assume pointers into this range are stack pointers is valuable for various programs (language runtimes, GCs, libcs, ...).

It would be nice if this were possible, and kind of requires that at least in practice, platform-specific code can be allowed to assume things run on a stack. Also, the existence of the `nostack` option in `asm!` seems dubious if there's no stack.

That said, IIRC C and C++ are careful not to promise anything about this. It's highly likely that we will not get many guarantees out of the codegen backend, as I'm not sure how you could promise much about this without limiting some very very important optimizations (SROA, store-load forwarding, etc), which are more important than the ability to do this[^2]

[^1]: Generally spawning a thread like this without using libc comes with the promise of "I will never call into libc again", but no_std code should *definitely* be allowed to do these sorts of things.

[^2]: In practice what I think you want is more of a "If this variable isn't in a register, and has its address taken, it will be on the stack", perhaps (which is typically trivial). But that would need more nailing down, I guess.

**Comment by JakobDegen** [comment]
Briefly discussed in backlog bonanza. We expect to not make this guarantee as it precludes a lot of important optimizations.

**Comment by JakobDegen** [comment]
@rust-lang/opsem we had originally discussed FCPing the "no" answer in the meeting, but looking at this issue, I'm not sure how useful that is. We don't really have a process for closing issues in this repo, so I thought it might be worth a little bit of discussion. The reason I can see for FCPing this is that a user (well, Niko) came to us with a question/request of the form "does this code work the way I'd like it to" and we're essentially saying "no." It might be useful to give team members the opportunity to object to that.

On the other hand, we're not really making any *decision* here. It's not as if we're specifying some behavior (or even specifying that some operation is UB). We're just saying that "the address of a local variable is located in address range R" is not a theorem that can be derived from our semantics as we envision them. If there was a concrete semantic question here, I would favor an FCP (even if the conclusion is "sorry, UB" or "sorry, nondet" or whatever).

I'm inclined to just close this as resolved in the negative without an FCP. Thoughts?

**Comment by RalfJung** [comment]
It would be nice to get some kind of deliverable out of this -- a piece of documentation summarizing the discussion. I am not sure what would be the best shape of that though. Some kind of FAQ? Or a general document on pointer identity?

**Comment by JakobDegen** [comment]
I've added an S-pending-documentation label with the description:

> Status: The issue is "resolved," but this resolution needs documentation

Keeping that issue open with that label sounds like a good idea. We probably don't have a good place to document this right now, but that's fine - we'll close this issue once we do

**Comment by chorman0773** [comment]
> I'm inclined to just close this as resolved in the negative without an FCP. Thoughts?

One use in running an FCP-close is to test that rfcbot functions on the repository :laughing:. 
From a procedure perspective, if we are indeed concretely answering the question here with "No, it's not guaranteed", it may be a good idea to establish as precedent that such negatives still need a sign-off and FCP, even if this particular thing isn't controversial in that answer. And if indeed documentation is being added, signing off *on* that documentation sounds like a good idea.

**Comment by Lokathor** [comment]
Yeah any answers that are "we know this for sure" should have an FCP, unless it's some duplicate of a previous FCP.
